CASSANDRA-8272 trunk Fix replica-side filtering

adelapena opened the pull request.

blerer commented on src/java/org/apache/cassandra/service/reads/DataResolver.java:
```diff
@@ -103,6 +103,156 @@ public PartitionIterator resolve()
+                                                                 i -> shortReadProtectedResponse(i, firstPhaseContext),
+                                                                 UnaryOperator.identity());
+
+        // Consume the fist phase partitions to populate the replica filtering protection with both those materialized
```
fist -> first

blerer commented on src/java/org/apache/cassandra/service/reads/DataResolver.java:
```diff
@@ -103,6 +103,156 @@ public PartitionIterator resolve()
+                                                                 UnaryOperator.identity());
+
+        // Consume the fist phase partitions to populate the replica filtering protection with both those materialized
+        // partitions and the primarey keys to be fetched.
```
primarey -> primary

jasonstack commented on src/java/org/apache/cassandra/db/partitions/PartitionIterators.java:
```diff
@@ -83,6 +83,18 @@ public static PartitionIterator singletonIterator(RowIterator iterator)
         return new SingletonPartitionIterator(iterator);
     }
 
+    public static void consume(PartitionIterator iterator)
```
do we need to close the given iterator?

adelapena replied:
Good catch. This `consume` method was present in 3.0 but it was removed in 3.11 because there weren't usages of it, so we are reintroducing it. The original method didn't close the iterator because the callers did so with try-with-resources blocks. I'm preserving that behaviour and closing the iterator in the caller (`DataResolver#resolveWithReplicaFilteringProtection`).

jasonstack commented on src/java/org/apache/cassandra/service/reads/ReplicaFilteringProtection.java:
```diff
@@ -0,0 +1,473 @@
+                            // ClusteringIndexNamesFilter we'll build from this later does not expect it), but the fact
+                            // we created a builder in the first place will act as a marker that the static row must be
+                            // fetched, even if no other rows are added for this partition.
+                            if (!isStatic)
```
if `static row` is the only outdated data, then `BTreeSet.Builder<Clustering> toFetch` would be empty..so `ClusteringIndexNamesFilter` will has no clusterings and query entire partition in `querySourceOnkey()`..

adelapena replied:
Right, I have changed to use limit=1 in there are no clusterings, similarly to what we do with in [`CQL3CasRequest`](https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/cql3/statements/CQL3CasRequest.java#L191-L200). I don't know if there is a more specific way of doing that, though.

jasonstack commented on src/java/org/apache/cassandra/service/reads/ReplicaFilteringProtection.java:
```diff
@@ -0,0 +1,473 @@
+        catch (UnavailableException e)
+        {
+            int blockFor = consistency.blockFor(keyspace);
+            throw UnavailableException.create(consistency, blockFor, e.alive);
```
should we use `blockFor - 1` instead of `e.alive`(0) for number of live replicas?

jasonstack commented on src/java/org/apache/cassandra/service/reads/DataResolver.java:
```diff
@@ -103,6 +103,156 @@ public PartitionIterator resolve()
+        // After reading the entire query results the protection helper should have cached all the partitions so we can
+        // clear the responses accumulator for the sake of memory usage, given that the second phase might take long if
+        // it needs to query replicas.
+        responses.clearUnsafe();
```
index range read won't trigger speculative read retry.. looks ok

jasonstack commented on src/java/org/apache/cassandra/service/reads/ReplicaFilteringProtection.java:
```diff
@@ -0,0 +1,473 @@
+ * <p>
+ * See CASSANDRA-8272 and CASSANDRA-8273 for further details.
+ */
+class ReplicaFilteringProtection<E extends Endpoints<E>>
```
I don't think replica filtering protection works with transient replica.. if a node is transient replica, repaired data will be removed on the node purposely to reduce storage requirement and causes unnecessary RFP..

maedhroz replied:
@jasonstack Is your worry just that we'll issue useless RFP requests to transient replicas that have already removed their versions of the relevant rows via incremental repair? Is that actually a correctness concern?

It seems like we should be able to ignore transient replicas most of the time anyway as long as we have enough full replicas to meet the original query's read CL...

jasonstack replied:
> It seems like we should be able to ignore transient replicas most of the time anyway as long as we have enough full replicas to meet the original query's read CL...

right.. it's not about correctness.. RFP should still work with TR, but have to send dummy protection request to transient replicas even if it's in-sync.

maedhroz replied:
Yeah. I don't have any good ideas on how we would avoid that, but this is already a somewhat degenerate case to begin with.

adelapena replied:
I guess we can just skip the requests for transient replicas, though I still have to think a bit more about the possible cases. It's worth mentioning that [secondary indexes are not supported in keyspaces with transient replicas](https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/cql3/statements/schema/CreateIndexStatement.java#L99-L100), although queries with `ALLOW FILTERING` seem to be supported.

maedhroz replied:
@adelapena Let me work through a couple examples in terms of whether or not we would expect problems with filtering queries...

**Example 1**: Say we have 3 replicas (2 full and 1 transient)

1.) `r1` (full) and `r3` (transient) receive a QUORUM write, but `r2` (full) misses it.
2.) `r2` (full) and `r3` (transient) receive a QUORUM delete, but `r1` (full) misses it.
2.) We issue a QUORUM read against `r1` and `r3` (or `r2`). Either way, the stale match from `r1` will be resolved out. _This case looks okay._

**Example 2**: Also 3 replicas (2 full and 1 transient)

1.) Write the value `c=1` at `ONE `with `timestamp=1` to `r1` (full) with `r2` (full) and `r3` (transient) down.
2.) Write the value `c=2` at `ONE` with `timestamp=2` to `r2` (full) with `r1` (full) and `r3` (transient) down.
3.) Write the value `c=3` at `ONE` with `timestamp=3` to `r3` (transient) with `r1` (full) and `r2` (full) down.
4.) Query for `c=1` at `ALL`. The only result filtering on the three replicas should produce is the stale match from `r1`. In this case, it seems like we would need RFP to make sure we get the latest row data _from the transient replica_ `r3` to produce the correct result, which is an empty one, at least _until_ `r1` or `r2` is incrementally repaired.

CC @jasonstack

adelapena replied:
@maedhroz So it seems we need to query transient replicas with RFP. Anyway I'm extending the dtests to cover transient replication in case we are missing something.

adelapena replied:
Just added [a couple of very basic dtests](https://github.com/apache/cassandra-dtest/pull/59/commits/c6df6e49be50347467dbf481b9b3176ca7e315c4) using transient replicas. They fail without RFP, and they pass with it.

adelapena commented:
Besides transient replicas, another pending matter is for what versions we want to apply the patch. Strictly this is a bug fix, but it has been known for so many time that it almost feels like an improvement, so I wonder what the fix version should be. Any thoughts?

blerer commented:
I would put it in 3.0, 3.11 and trunk. It is a bug fix. :-)
I would ignore 2.2 because it would be soon phased out.

adelapena committed:
Fix replica-side filtering returning stale data with CL > 1

patch by Andres de la Peña; reviewed by Benjamin Lerer, Caleb Rackliffe and ZhaoYang for CASSANDRA-8272

adelapena commented:
> I would put it in 3.0, 3.11 and trunk. It is a bug fix. :-)
I would ignore 2.2 because it would be soon phased out.

Makes sense. I have created a new PR for 3.0, #603. Dtests are slightly modified to skip those using features that are not supported on 3.0 (indexes on static columns and filtering on partition key parts). I'm running CI, I'll post the results here.

adelapena commented:
CI results:

| branch | CircleCI | cassandra-ci utest | cassandra-ci dtest |
|--------|---------|--------------------|--------------------|
|3.0|[44](https://app.circleci.com/pipelines/github/adelapena/cassandra/44/workflows/15ab85cc-6c8c-4625-b98a-1337da433e91)|[109](https://ci-cassandra.apache.org/view/all/job/Cassandra-devbranch-test/109/)|[135](https://ci-cassandra.apache.org/view/all/job/Cassandra-devbranch-dtest/135/)|
|3.11|[45](https://app.circleci.com/pipelines/github/adelapena/cassandra/45/workflows/52db2155-ad14-4daf-809e-e812e3be0be6)|[110](https://ci-cassandra.apache.org/view/all/job/Cassandra-devbranch-test/110/)|[136](https://ci-cassandra.apache.org/view/all/job/Cassandra-devbranch-dtest/136/)|
|trunk|[46](https://app.circleci.com/pipelines/github/adelapena/cassandra/46/workflows/6d792fe7-4a45-4962-8ef9-869ef9d9dd72)|[111](https://ci-cassandra.apache.org/view/all/job/Cassandra-devbranch-test/111/)|[137](https://ci-cassandra.apache.org/view/all/job/Cassandra-devbranch-dtest/137/)|

Utests look good. The most suspicious dtest failure is `TestRepair.test_dead_sync_initiator` on trunk. That doesn't seem related to the changes to me, though I'm digging into it.

adelapena committed:
Merge branch '8272-3.0' into 8272-3.11

# Conflicts:
#	src/java/org/apache/cassandra/db/filter/RowFilter.java

adelapena committed:
Merge branch '8272-3.11' into 8272-trunk

# Conflicts:
#	src/java/org/apache/cassandra/db/PartitionRangeReadCommand.java
#	src/java/org/apache/cassandra/db/ReadCommand.java
#	src/java/org/apache/cassandra/db/SinglePartitionReadCommand.java
#	src/java/org/apache/cassandra/db/filter/RowFilter.java
#	src/java/org/apache/cassandra/db/partitions/UnfilteredPartitionIterators.java
#	src/java/org/apache/cassandra/metrics/TableMetrics.java
#	src/java/org/apache/cassandra/service/DataResolver.java
#	src/java/org/apache/cassandra/utils/concurrent/Accumulator.java

adelapena committed:
CircleCI

adelapena committed:
Restore accidentally removed @SupressWarnings tag

maedhroz commented:
@adelapena So far, some local runs of `test_dead_sync_initiator` against trunk look pretty stable, but I'm going to try with your trunk-based branches...

jasonstack commented:
The failure should go away after rebase dtest.

maedhroz commented:
@adelapena For what it's worth, I also wasn't ablate reproduce any failure around `test_dead_sync_initiator` w/ your branches.

adelapena commented:
These are the CI results after rebasing everything:

| branch | CircleCI | cassandra-ci utest | cassandra-ci dtest |
|--------|---------|--------------------|--------------------|
|3.0|[48](https://app.circleci.com/pipelines/github/adelapena/cassandra/48/workflows/fcb8e1ab-545a-47f2-9432-03b783953f88)|[113](https://ci-cassandra.apache.org/view/all/job/Cassandra-devbranch-test/113/)|[141](https://ci-cassandra.apache.org/view/all/job/Cassandra-devbranch-dtest/141/)|
|3.11|[49](https://app.circleci.com/pipelines/github/adelapena/cassandra/49/workflows/894eb560-45c9-44a9-b48b-6ed2f47b2298)|[114](https://ci-cassandra.apache.org/view/all/job/Cassandra-devbranch-test/114/)|[142](https://ci-cassandra.apache.org/view/all/job/Cassandra-devbranch-dtest/142/)|
|trunk|[50](https://app.circleci.com/pipelines/github/adelapena/cassandra/50/workflows/0fa01053-2b96-4880-afb2-d8fe020b8848)|[115](https://ci-cassandra.apache.org/view/all/job/Cassandra-devbranch-test/115/)|[143](https://ci-cassandra.apache.org/view/all/job/Cassandra-devbranch-dtest/143/)|

It seems that `test_dead_sync_initiator ` still fails for some runs after rebasing ([here](https://ci-cassandra.apache.org/view/all/job/Cassandra-devbranch-dtest/143/testReport/junit/dtest.repair_tests.repair_test/TestRepair/test_dead_sync_initiator/) and [here](https://app.circleci.com/pipelines/github/adelapena/cassandra/50/workflows/0fa01053-2b96-4880-afb2-d8fe020b8848/jobs/409)), although I don't think it's related to the changes introduced by this PR. I haven't been able to reproduce the failure locally.

There is also [a failure in `CustomIndexTest.testFailing2iFlush`](https://app.circleci.com/pipelines/github/adelapena/cassandra/50/workflows/0fa01053-2b96-4880-afb2-d8fe020b8848/jobs/404) in the CircleCI run for 8272-trunk. Multiplexer runs show that this test is slightly flaky in 8272-trunk, whereas it looks stable in trunk:
| branch | run | iterations|failures |
|--------|-----|----------|-|
|trunk|[653](https://jenkins-cassandra.datastax.lan/job/parameterized-testall/653/)|200|0|
|trunk|[654](https://jenkins-cassandra.datastax.lan/job/parameterized-testall/654/)|200|0|
|8272-trunk|[652](https://jenkins-cassandra.datastax.lan/job/parameterized-testall/652/)|100|1|
|8272-trunk|[655](https://jenkins-cassandra.datastax.lan/job/parameterized-testall/655/)|200|2|

That test uses indexes but not in through read path nor with multiple nodes, so I don't know how the failure can be caused by this PR. I'll try to reproduce it locally, and make sure it doesn't really affect trunk.

@jasonstack other than that, do you agree to applying replica filtering protection with transient replicas?

jasonstack commented:
> do you agree to applying replica filtering protection with transient replicas?

I don't see any correctness issue.  +1

jasonstack commented:
> It seems that test_dead_sync_initiator still fails for some runs after rebasing (here and here), although I don't think it's related to the changes introduced by this PR. I haven't been able to reproduce the failure locally.

I don't think this is related to this patch.

I believe the invalid checksum is related to zero-copy-streaming handling channel close gracefully and then leaving files incomplete. I will create a ticket on this.

adelapena commented:
I have run `CustomIndexTest#testFailing2iFlush` 1000 times without failures neither in trunk nor 8272-trunk:
| branch | run | iterations|failures |
|--------|-----|-----------|---------|
|trunk|[659](https://jenkins-cassandra.datastax.lan/job/parameterized-testall/659/)|500|0|
|trunk|[662](https://jenkins-cassandra.datastax.lan/job/parameterized-testall/662/)|500|0|
|8272-trunk|[660](https://jenkins-cassandra.datastax.lan/job/parameterized-testall/660/)|500|0|
|8272-trunk|[661](https://jenkins-cassandra.datastax.lan/job/parameterized-testall/661/)|500|0|

It hasn't failed in [an additional run of CircleCI](https://app.circleci.com/pipelines/github/adelapena/cassandra/50/workflows/6d9acf4a-d4ce-4604-a64b-38887c6160ab), where there is only a failure of the unrelated `test_dead_sync_initiator `, and I haven't been to reproduce it locally.

I don't know what made `testFailing2iFlush` fail in both CircleCI and the multiplexer, but it seems that there is no way to reproduce it, and also the test doesn't seem related to what we are changing here. Thus, I think we are probably ready to commit. WDYT? Should I commit?

blerer commented:
Let's commit it.

adelapena commented:
Done, thanks for the reviews
